code/game/objects/structures/electricchair.dm

```diff
@@ -1,52 +1,132 @@
 
 /obj/structure/chair/e_chair/attackby(obj/item/W, mob/user, params)
 	if(W.tool_behaviour == TOOL_WRENCH)
 		var/obj/structure/chair/C = new /obj/structure/chair(loc)
 		W.play_tool_sound(src)
 		C.setDir(dir)
-		part.forceMove(loc)
-		part.master = null
-		part = null
+		stored_kit.forceMove(loc)
+		stored_kit.master = null
+		stored_kit = null
 		qdel(src)
 
-/obj/structure/chair/e_chair/proc/shock()
-	if(last_time + 50 > world.time)
-		return
-	last_time = world.time
 
-	// special power handling
-	var/area/A = get_area(src)
-	if(!isarea(A))
-		return
-	if(!A.powered(AREA_USAGE_EQUIP))
+/datum/component/electrified_chair
+	///the chair that we are attached to
+	var/obj/structure/chair/parent_chair
+	///the shock kit attached to parent_chair
+	var/obj/item/assembly/shock_kit/used_shock_kit
+	///the mob buckled to parent_chair, if any
+	var/mob/living/guinea_pig
+	///the guinea pig can only be shocked if this reaches 5 seconds
+	var/time_since_last_shock = 0
+	///this is casted to the overlay we put on parent_chair
+	var/mutable_appearance/shock_helmet_overlay
+	///the original name of parent_chair
+	var/original_name
+
+/datum/component/electrified_chair/Initialize(obj/item/assembly/shock_kit/input_shock_kit)
+	if(!istype(parent, /obj/structure/chair) || !istype(input_shock_kit, /obj/item/assembly/shock_kit))
+		return COMPONENT_INCOMPATIBLE
+	parent_chair = parent
+	used_shock_kit = input_shock_kit
+	RegisterSignal(parent_chair, COMSIG_PARENT_PREQDELETED, .proc/delete_self)
+	RegisterSignal(used_shock_kit, COMSIG_PARENT_PREQDELETED, .proc/delete_self)
+
+	RegisterSignal(parent_chair, COMSIG_MOVABLE_BUCKLE, .proc/on_buckle)
+	RegisterSignal(parent_chair, COMSIG_ATOM_EXIT, .proc/check_shock_kit)
+	shock_helmet_overlay = image('icons/obj/chairs.dmi', "echair_over", OBJ_LAYER)
+	parent_chair.add_overlay(shock_helmet_overlay)
+	original_name = parent_chair.name
+	parent_chair.name = "electrified [original_name]"
+
+	if(parent_chair.has_buckled_mobs())
+		for(var/m in parent_chair.buckled_mobs)
+			var/mob/living/possible_guinea_pig = m
+			if(on_buckle(src, possible_guinea_pig, FALSE))
+				break
+
+/datum/component/electrified_chair/UnregisterFromParent()
+	parent_chair.cut_overlay(list(shock_helmet_overlay))
+	parent_chair.name = original_name
+	if(parent_chair)
+		UnregisterSignal(parent_chair, list(COMSIG_PARENT_PREQDELETED, COMSIG_MOVABLE_BUCKLE, COMSIG_MOVABLE_UNBUCKLE, COMSIG_ATOM_EXIT))
+	if(used_shock_kit)
+		UnregisterSignal(used_shock_kit, list(COMSIG_PARENT_PREQDELETED))
+	if(guinea_pig)
+		UnregisterSignal(guinea_pig, list(COMSIG_PARENT_PREQDELETED))
+	parent_chair = null
+	used_shock_kit = null
+	guinea_pig  = null
+	STOP_PROCESSING(SSprocessing, src)
+
+/datum/component/electrified_chair/proc/delete_self()
+	SIGNAL_HANDLER
+	qdel(src)
+
+///checks if the shock kit connected to parent is still there and unregisters if it isnt
+/datum/component/electrified_chair/proc/check_shock_kit(datum/source, atom/movable/AM, atom/newLoc)
```

Comment:
Better variable name.

Signal handlers don't have the same caveat as re-implementations, as you can't use named parameters for them anyway, as far as I know.

Reply:
proc no longer exists anyways